File: helparty/src/main/java/com/hamryt/helparty/service/gym/GymServiceImpl.java
package com.hamryt.helparty.service.gym;

import com.hamryt.helparty.dto.UserType;
import com.hamryt.helparty.dto.gym.GymDTO;
import com.hamryt.helparty.dto.gym.request.SignUpGymRequest;
import com.hamryt.helparty.dto.gym.request.UpdateGymRequest;
import com.hamryt.helparty.dto.gym.response.SignUpGymResponse;
import com.hamryt.helparty.dto.gym.response.UpdateGymResponse;
import com.hamryt.helparty.exception.gym.GymDeleteFailedException;
import com.hamryt.helparty.exception.gym.GymNotFoundByIdException;
import com.hamryt.helparty.exception.gym.GymNotFoundException;
import com.hamryt.helparty.exception.gym.InsertGymFailedExcetpion;
import com.hamryt.helparty.exception.user.DoesNotMatchUserType;
import com.hamryt.helparty.exception.user.EmailExistedException;
import com.hamryt.helparty.exception.user.UpdateFailedException;
import com.hamryt.helparty.mapper.GymMapper;
import com.hamryt.helparty.service.session.Encryptor;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Service
public class GymServiceImpl implements GymService {
    
    private final Encryptor encryptor;
    private final GymMapper gymMapper;
    
    @Transactional
    public SignUpGymResponse insertGym(SignUpGymRequest signUpGymRequest) {
        
        if (signUpGymRequest.getUserType() != UserType.GYM) {
            throw new DoesNotMatchUserType(UserType.GYM);
        }
        
        if (isExistsEmail(signUpGymRequest.getEmail())) {
            throw new EmailExistedException(signUpGymRequest.getEmail());
        }
        
        String encodedpassword = encryptor.encrypt(signUpGymRequest.getPassword());
        
        GymDTO newGym = GymDTO.builder()
            .email(signUpGymRequest.getEmail())
            .gymName(signUpGymRequest.getGymName())
            .password(encodedpassword)
            .phoneNumber(signUpGymRequest.getPhoneNumber())
            .addressCode(signUpGymRequest.getAddressCode())
            .addressDetail(signUpGymRequest.getAddressDetail())
            .userType(signUpGymRequest.getUserType())
            .build();
        
        if (gymMapper.insertGym(newGym) != 1) {
            throw new InsertGymFailedExcetpion();
        }
        
        return SignUpGymResponse.of(newGym);
    }
    
    @Transactional(readOnly = true)
    public GymDTO findGymByEmailAndPassword(String email, String password) {
        return Optional.ofNullable(gymMapper.findGymByEmailAndPassword(email, password))
            .orElseThrow(() -> new GymNotFoundException(email));
    }
    
    @Transactional
    public UpdateGymResponse updateGym(Long id, UpdateGymRequest updateGymRequest) {
        
        String encodedPassword = encryptor.encrypt(updateGymRequest.getPassword());
        
        UpdateGymResponse updateGymResponse = UpdateGymResponse
            .of(id, encodedPassword, updateGymRequest);
        
        if (gymMapper.updateGym(updateGymResponse) != 1) {
            throw new UpdateFailedException(id);
        }
        return updateGymResponse;
    }
    
    @Transactional(readOnly = true)
    public String findGymEmailById(Long id) {
        return Optional.ofNullable(gymMapper.findGymEmailById(id))
            .orElseThrow(() -> new GymNotFoundByIdException(id));
    }
    
    @Transactional
    public void deleteGym(Long id) {
        if (gymMapper.deleteGymById(id) != 1) {
            throw new GymDeleteFailedException(id);
        }
    }
    
    @Transactional(readOnly = true)
    public boolean isExistsEmail(String email) {
        return gymMapper.isExistsEmail(email);
    }
}

File: helparty/src/main/java/com/hamryt/helparty/service/gym/GymServiceImpl.kt
package com.hamryt.helparty.service.gym

import com.hamryt.helparty.dto.gym.GymDTO
import com.hamryt.helparty.dto.gym.request.SignUpGymRequest
import com.hamryt.helparty.dto.gym.request.UpdateGymRequest
import com.hamryt.helparty.dto.gym.response.SignUpGymResponse
import com.hamryt.helparty.dto.gym.response.UpdateGymResponse
import com.hamryt.helparty.exception.gym.GymDeleteFailedException
import com.hamryt.helparty.exception.gym.GymNotFoundException
import com.hamryt.helparty.exception.gym.InsertGymFailedException
import com.hamryt.helparty.exception.user.EmailExistedException
import com.hamryt.helparty.exception.user.UpdateFailedException
import com.hamryt.helparty.mapper.GymMapper
import com.hamryt.helparty.service.session.Encryptor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import javax.validation.constraints.NotEmpty

@Service
class GymServiceImpl(

        private val encryptor: Encryptor,

        private val gymMapper: GymMapper

) : GymService {

    override fun insertGym(resource: SignUpGymRequest): SignUpGymResponse {
        resource.userType.validEqualUserType("GYM")

        if(isExistsEmail(resource.email)){
            throw EmailExistedException(resource.email)
        }

        var encodedpassword = encryptor.encrypt(resource.password)

        var newGym = GymDTO.create(
                null,
                resource.email,
                resource.gymName,
                encodedpassword,
                resource.phoneNumber,
                resource.addressCode,
                resource.addressDetail,
                resource.userType
        )

        if(gymMapper.insertGym(newGym) != 1){throw InsertGymFailedException()}

        return SignUpGymResponse.of(newGym)
    }

    @Transactional(readOnly = true)
    override fun findGymByEmailAndPassword(email: String, password: String): GymDTO {
        return gymMapper.findGymByEmailAndPassword(email, password)?: throw GymNotFoundException(email)
    }

    override fun updateGym(loginId: Long, updateGymRequest: UpdateGymRequest): UpdateGymResponse {
        var encodedPassword = encryptor.encrypt(updateGymRequest.password)

        var updateGymResponse = UpdateGymResponse.of(loginId, encodedPassword, updateGymRequest)

        if(gymMapper.updateGym(updateGymResponse)!=1){
            throw UpdateFailedException(loginId);
        }

        return updateGymResponse
    }

    override fun deleteGym(loginId: Long) {
        if(gymMapper.deleteGymById(loginId) != 1){
            throw GymDeleteFailedException(loginId)
        }
    }

    @Transactional(readOnly = true)
    private fun isExistsEmail(email: @NotEmpty String) = gymMapper.isExistsEmail(email)

}